java/src/main/java/com/genexus/db/driver/GXPreparedStatement.java

```diff
@@ -934,89 +933,100 @@ public void setGXDbFileURI(int index, String fileName, String blobPath, int leng
-				fileUri = GXDbFile.generateUri(fileName, !GXDbFile.hasToken(fileName), true);
-			else
-			{
-				if (blobPath.trim().length() != 0)
-				{
-					blobPath = com.genexus.GXutil.cutUploadPrefix(blobPath);
-					File file = new File(blobPath);
-					fileUri = GXDbFile.generateUri(file.getName(), !GXDbFile.hasToken(blobPath), true);
+    public void setGXDbFileURI(int index, String fileName, String blobPath, int length, String tableName, String fieldName) throws SQLException {
+
+		fileName = fileName.trim();
+		blobPath = blobPath.trim();
+
+		//EMPTY BLOB
+    	if (blobPath == null || blobPath.trim().length() == 0) {
+			setVarchar(index, fileName, length, false);
+			return;
+		}
+
+		String fileUri = "";
+    	String multimediaTemporalPath = com.genexus.Preferences.getDefaultPreferences().getMultimediaPath();
+
+		if (fileName.length() > 0 && !(blobPath.startsWith(multimediaTemporalPath) && GXutil.isAbsoluteURL(fileName))) {
+			fileUri = GXDbFile.generateUri(fileName, !GXDbFile.hasToken(fileName), true);
+		}
+		else if (blobPath.trim().length() > 0)
+		{
+			blobPath = com.genexus.GXutil.cutUploadPrefix(blobPath);
+			File file = new File(blobPath);
+			fileUri = GXDbFile.generateUri(file.getName(), !GXDbFile.hasToken(blobPath), true);
+		}
+
+		ExternalProvider storageProvider = Application.getExternalProvider();
+		boolean externalStorageEnabled = storageProvider != null;
+		boolean attInExternalStorage = externalStorageEnabled && (tableName != null && fieldName != null); //Should improve this condition in order to not depend on tableName and FieldName.
+
+		if (!attInExternalStorage) {
+			setVarchar(index, fileUri, length, false);
+			return;
+		}
+
+		//External Storage is ENABLED. We have the following cases:
+		// - 1. WebUpload: The URL is an External Storage URL in the Private Temp Storage Folder.
+		// - 2. External URL: An absolute URL (outside External Storage).
+		// - 3. Transaction Update Mode (nothing changed): An Storage URL that is already in External Storage (except in private), should not be uploaded nor copied.
+		// - 4. Upload Resource from local drive to External Storage: Ex: Image.FromImage(ActionDelete)
+
+		String storageObjectName = ExternalProviderCommon.getProviderObjectName(storageProvider, blobPath);
+		boolean resourceAlreadyOnStorage = storageObjectName != null;
+		ResourceAccessControlList defaultAcl = ResourceAccessControlList.Default;
+
+		String storageTargetObjectName = blobPath;
+		boolean isPrivateTempUpload = storageTargetObjectName.startsWith(Application.getClientPreferences().getTMPMEDIA_DIR());
+		String folder = Application.getGXServices().get(GXServices.STORAGE_SERVICE).getProperties().get("FOLDER_NAME");
+		if (!resourceAlreadyOnStorage && (fileName == blobPath || GXutil.isAbsoluteURL(fileName))) {
+			// - 2. External URL: An absolute URL (outside External Storage).
+			try {
+				InputStream is = new URL(fileName).openStream();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *is*. Currently, there are execution paths that do not contain closure statements, for example, when upload throws an exception. Either a) close *is* in a try-finally block or b) close the resource by declaring *is* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).